### ISSUE TITLE
Move module-info back to the main root?

### ISSUE THREAD

**Issue by cpovirk** [jspecify/jspecify#484]
Currently, it's under the 9+ section of the mrjar. That's fine, but I wonder if we'd be better off getting modular-style Javadoc output, as we might if we moved it to the root? (Such output might either help or hurt people who are trying to link to us. Maybe it is part of the story with https://github.com/maxcellent/javadoc.io/issues/182, but I haven't tried to understand that....)

**Comment by cpovirk** [comment]
If the only reason to care is Javadoc, we might also be able to just run Javadoc differently.

My feeling has been that this isn't too important. But if it changes the URLs of our Javadoc output, then we might want to do it sooner rather than later (or decide to _never_ do it, which may or may not be swimming upstream). Or we could insert our own redirects from the old locations to the new at any point.

**Comment by cpovirk** [comment]
I have discovered an extremely trivial reason to put `module-info` in the root: It makes [`--describe-module`](https://stackoverflow.com/a/46617956/28465) slightly easier to use (but kudos to it for the helpful error message!):

```
$ jar --describe-module --file j2objc-annotations-3.0.0.jar
releases: 9

No root module descriptor, specify --release

$ jar --describe-module --file j2objc-annotations-3.0.0.jar --release 9
releases: 9

com.google.j2objc.annotations@3.0.0 jar:file:///tmp/jobjc_release.SF8NzIp/j2objc/annotations/target/j2objc-annotations-3.0.0.jar/!META-INF/versions/9/module-info.class open
exports com.google.j2objc.annotations
requires java.base
```

Presumably the original command would Just Work if the `module-info` were in the root.

Anyway, I don't think that should have much effect on a decision here. I'm still not sure if either way is significantly better than the other.

**Comment by cpovirk** [comment]
I just came across https://github.com/jspecify/jspecify/issues/245#issuecomment-1122425671 again, which suggests that the Gradle "release" setting might not fully behavior like javac's `--release` flag, at least when the APIs differ across versions? We concluded there that we should just make the APIs _not_ differ across versions. So the question now would be whether putting `module-info` under the Java 9 mrjar root counts as an "API difference" for that purpose. (Or we could still flip the question around and ask why _not_ to put the `module-info` under the main root. If we're concerned about Javadoc (as discussed above), we can always run Javadoc ourselves when we want to, as discussed in https://github.com/jspecify/jspecify/pull/527#issuecomment-2173586291.)

**Comment by agentgt** [comment]
@cpovirk @kevinb9n Since you guys are talking about module-info and placement another consideration besides `META-INF/releases` is that `module-info` has the version of javac used to compile it (at least with maven)! When I say version I mean including patch version.

So if you compile a `module-info` with javac `17.0.11` the module-info will have `17.0.11` embedded init.

https://github.com/jvm-repo-rebuild/reproducible-central/issues/113

I'm not sure if Gradle has fixed that but I believe it is a problem even with the JDK but has been fixed in 22 and I believe Maven.

To fix it for my case I force the build to use an exact version of javac for reproducibility. The other option is to strip the info of which I'm not sure what tools are out there that do that especially gradle wise.

Anyway sorry if I wasted time if you guys already know this.

**Comment by translatenix** [comment]
My JPMS module uses JSpecify.
I get the following warning every time I build my Javadoc (which links to JSpecify's Javadoc):

```
warning: The code being documented uses modules but the packages defined in https://jspecify.dev/docs/api/ are in the unnamed module.
```

**Comment by cpovirk** [comment]
Interesting, thanks. Are you able to link to the project or share a smaller project that reproduces the warning? Alternatively, have you been able to confirm that moving the module info resolves the warning?

**Comment by translatenix** [comment]
Did you change minipack's `javadoc` task to point to your locally built javadoc?

https://github.com/translatenix/minipack/blob/main/minipack-core/build.gradle.kts#L56

**Comment by cpovirk** [comment]
Nope, sorry about that! Let me see....

**Comment by cpovirk** [comment]
Note to self: To actually build Javadoc successfully, you need to set the `JAVA_VERSION` environment variable, as in:

```
JAVA_VERSION=23 JAVA_HOME=$HOME/jdk-23-ea+22 ./gradlew javadoc
```

**Comment by cpovirk** [comment]
Err, not that _that_ command works for me _at head_, on account of:

```
src/main/java/org/jspecify/annotations/NullMarked.java:31: error: reference not found
 * javax.lang.model.element.Element#getEnclosedElements() enclosed} within it are <b>null-marked
   ^
src/main/java/org/jspecify/annotations/NullUnmarked.java:30: error: reference not found
 * javax.lang.model.element.Element#getEnclosedElements() enclosed} within it is <b>null-unmarked
   ^
```

But it apparently works fine in GitHub Actions, as shown by the working Javadoc at https://jspecify.dev/docs/api/org/jspecify/annotations/NullMarked.html. I'm going to try with an actual release of 23.

**Comment by cpovirk** [comment]
Wait, no, sorry, it _does_ work at _head_, but it does _not_ work with my hacked-up changes! Those changes include not only a root `module-info` but also a change of `sourceCompatibility` and `targetCompatibility` to 9. So presumably we need to do _something_ like that to get Gradle and/or javadoc into a sort of "modules mode." (Presumably we can adjust the settings for the Javadoc run but not the main build.) Then (well, after we fix the `javax.lang.model.element` issue) we'd get the modular Javadoc output that you're looking for (`element-list`: `module:org.jspecify`, `org.jspecify.annotations`). That includes changing the URL schema, breaking any links. I'm not sure there's [a way to avoid that anymore](https://bugs.openjdk.org/browse/JDK-8215580), but we could set up redirects.

That is more than I am going to bite off now, but I'm glad to understand it better. Thanks for the report and the correction earlier. I've filed https://github.com/jspecify/jspecify/issues/712 to keep track of this. I _suspect_ it's independent of where the `module-info` ends up in the jar, but we'll see.

**Comment by translatenix** [comment]
Thanks for filing #712.
FYI, to temporarily suppress Javadoc errors, add this line to your `build.gradle`:
```
javadoc {
    options.encoding = 'UTF-8'
    options.addStringOption('Xdoclint:none', '-quiet')
}
```

**Comment by ben-manes** [comment]
An alternative workaround is to use `addStringOption("-link-modularity-mismatch", "info")` in Java 18 or above which will suppress these warnings. You can also specify the javadoc tool version to be later than the release version, e.g. run the JDK 23's javadoc while releasing for JDK 8, so that you get the latest javadoc output that links to the prior JDK release classes. You can review my project's [configuration](https://github.com/ben-manes/caffeine/blob/2542ca8ca22fe4c30b62ee569ee71436fc44a37e/gradle/plugins/src/main/kotlin/lifecycle/java-library.caffeine.gradle.kts#L91-L114) if helpful.

**Comment by cpovirk** [comment]
It seems believable (but far from guaranteed) that a root-directory `module-info` would have avoided the problem in https://github.com/hierynomus/smbj/issues/865 \[edit: and https://github.com/bebsworthy/pocket_agent/commit/3b2d71559e723231241484c93da29d2763b9e168 and https://github.com/customerio/customerio-flutter/issues/304 and https://github.com/CybotTM/kidsync/commit/6e93fc7b51c0ed6e76afbaa2b75c2d1b22bd2cc5 and https://github.com/bmander/thecityandthebike/commit/1364625da50eaee7094b6affe5bf218ef56f519c and https://github.com/Loken75/password-manager/commit/31601fb493c5abbe337f48d5ebdbf669670110ca and https://github.com/liekzejaws/rushlight/pull/2 and https://github.com/CallMeTechie/OriDev/pull/161 and https://github.com/dappros/ethora-sdk-android/commit/a965cdbc575ecc51d8a8ef29766e6f2b76dc1187 + https://github.com/dappros/ethora-sdk-android/commit/d120f91c59040e16f904b702e57a956047ec55cc and https://github.com/kiloiam/BiliPai-Article/commit/d5a51ef4abfb41c241e27115f18cbdd165d5369d and https://github.com/SaadRimeh/SaloumApp/commit/45b99a73d7dff97a78fb2b1b99b550ef958679ee and https://github.com/nganlinh4/screen-goated-toolbox/commit/824e146362e35fb2a94eae38a080f6695dcdcc15 and https://github.com/szerzogabor/mesha/pull/284 and https://github.com/IvoryCobra-VC/lidseeker/pull/37 and https://github.com/1c-syntax/intellij-language-1c-bsl/commit/31803679ee2e0340795c8956f7823e811cc59e37 and https://github.com/AbakoDolla/SXB-VPN/commit/61e3c1b6b2a90e7859098f70a955f640790afcfb and https://github.com/workingfivetonine/TimetoPayApp/commit/1ea6aaeee158ac972822438655774b862cd6a52c and https://github.com/boxcreate/boxlore/commit/fc045006408b29f88e1cc84ab183567d1947068a#diff-cd364d6c7d2884761543189df41b262b804d4cb5c61b4e8f35f9656892420267R48 and https://github.com/Metalrott420/intermingled-/commit/16dcce9d813bb97a9bced8993228543402b09d2d and likely https://github.com/convivae/SystemUI-Gradle/commit/0766a30ba88173d11b24a728a85b9f94b27e0f14#diff-967f76ee7cddd1107a40f46aedd0911a39e3ac782daaeb542f09c67c3361491fR44 and https://github.com/Anon5T4R/LocalKeysAndroid/commit/6936056af3aca38f6b900b2c5c17382f686aab9e and https://github.com/kaelen-phoenix/yo-boot/commit/500b20d1b7116dbe74284d29dbea278e27b0ba8d and https://github.com/markmtm/synccast/commit/d315258e1dae1ba34717287620b84f24c48d5fa1 and https://github.com/dimxyp/openssh-android/commit/a67da0b144fc8baacb3f98d8383a91765c9ce95a and https://github.com/xnexgamerx-byte/Social-Hub44/commit/b22a3d7ea3c28bfe944bd729f9a406067bd0d1df and https://github.com/maragung/EclipseSSH/commit/0b3cb3d7195f447e2d4fd1ce21bad6457c34fd1e and https://github.com/ReSerendipity/DraftPeek/commit/c290948e32478bc803dae702a9c84dd41dbb26e4 and https://github.com/Gooseberry111/Between-Us/commit/3e5dc6a10b908280399230455af85881fb5dfe83, too\]: Perhaps some tool understands that it may see multiple "versions" of `META-INF/MANIFEST.MF` but not `META-INF/versions/9/OSGI-INF/MANIFEST.MF`, and we could eliminate the latter by not having an additional root? But I'm guessing here. (And even if I'm right, the ideal solution would be for the tool to be fixed.)

\[edit: and how about the OSGi part of google/error-prone#4971?\]

\[edit: And avoiding multi-release jars might be nice in general because of occasional other incomplete support for them, if I'm reading https://github.com/PIsberg/nanometer/issues/17 right.\]

**Comment by cowwoc** [comment]
Maybe I've missed something but...

1. From a specification point-of-view, doesn't a JAR file require all classes to be compiled under the same JDK version, and classes compiled by a different JDK version to be placed under `META-INF/versions`? Consider that line items in `META-INF/MANIFEST.MF` make that assumption.
2. You should absolutely be able to compile your code using one JDK version and your Javadoc using a different one. I've had to do something similar for some of my projects due to bugs in older versions of the `javadoc` tool.

So to recap: can't you compile your base classes using JDK 8, your `module-info.java` class using JDK 9+, and your Javadoc using JDK 25? That sounds like the safest route to me.

**Comment by cpovirk** [comment]
I wonder if this change was working around our current `module-info` setup:

https://github.com/sparrowlogic/whisper4j/commit/9c3f31abca0866dc06826e147cf683b4617225c2

**Comment by sormuras** [comment]
As a side-note: JUnit 5's multi-release artifacts ship with `module-info.class`es in their modular JAR file roots since version [5.5.0](https://docs.junit.org/5.5.0/release-notes/) (June 30, 2019) although base classes target Java 8. We never received any issue reports in that regard.

**Comment by cpovirk** [comment]
> Since you guys are talking about module-info and placement another consideration besides `META-INF/releases` is that `module-info` has the version of javac used to compile it (at least with maven)! When I say version I mean including patch version.
> 
> So if you compile a `module-info` with javac `17.0.11` the module-info will have `17.0.11` embedded init.
> 
> [jvm-repo-rebuild/reproducible-central#113](https://github.com/jvm-repo-rebuild/reproducible-central/issues/113)
> 
> I'm not sure if Gradle has fixed that but I believe it is a problem even with the JDK but has been fixed in 22 and I believe Maven.

While going over things with a fine-tooth comb in light of https://github.com/jspecify/jspecify/issues/867, I noticed that (besides the intentional addition of the _JSpecify_ version for https://github.com/jspecify/jspecify/issues/746), we've picked up a "9" (the _new_ "#11" value below), apparently in our `requires java.base` line. (I hadn't noticed because it isn't reported by `jar --describe-module`):

```patch
@@ -17,21 +17,23 @@
    #5 = Utf8               Module
    #6 = Unknown            #7             // "org.jspecify"
    #7 = Utf8               org.jspecify
-   #8 = Unknown            #9             // "java.base"
-   #9 = Utf8               java.base
-  #10 = Unknown            #11            // org/jspecify/annotations
-  #11 = Utf8               org/jspecify/annotations
+   #8 = Utf8               1.0.1
+   #9 = Unknown            #10            // "java.base"
+  #10 = Utf8               java.base
+  #11 = Utf8               9
+  #12 = Unknown            #13            // org/jspecify/annotations
+  #13 = Utf8               org/jspecify/annotations
 {
 }
 SourceFile: "module-info.java"
 Module:
   #6,0                                    // "org.jspecify"
-  #0
+  #8                                      // 1.0.1
   1                                       // requires
-    #8,8000                                 // "java.base" ACC_MANDATED
-    #0
+    #9,8000                                 // "java.base" ACC_MANDATED
+    #11                                     // 9
   1                                       // exports
-    #10,0                                   // org/jspecify/annotations
+    #12,0                                   // org/jspecify/annotations
   0                                       // opens
   0                                       // uses
   0                                       // provides
```

Based on my description, Gemini suggest this is the [`compiledVersion()`](https://docs.oracle.com/en/java/javase/26/docs/api/java.base/java/lang/module/ModuleDescriptor.Requires.html#compiledVersion()).

The value of 9 seems reasonable (and probably mostly inconsequential to begin with, at least as long as it's predictable!): It's the `--release` value that we pass, and we do so because it's (obviously) the lowest version to support modules.

This is another reason to be happy that we [force a specific, new version of javac](https://github.com/jspecify/jspecify/pull/818) so that we benefit from [the mentioned fix](https://bugs.openjdk.org/browse/JDK-8318913). (And it might be another reason to be happy to probably not have a strict need to [further touch our build system much](https://github.com/jspecify/jspecify/issues/865) at the moment.)
